File: src/datahub/databus/databus-pipe-extractor/src/main/java/com/tencent/bk/base/datahub/databus/pipe/Context.java
/*
 * Tencent is pleased to support the open source community by making BK-BASE 蓝鲸基础平台 available.
 *
 * Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * BK-BASE 蓝鲸基础平台 is licensed under the MIT License.
 *
 * License for BK-BASE 蓝鲸基础平台:
 * --------------------------------------------------------------------
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
 * NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.tencent.bk.base.datahub.databus.pipe;

import com.tencent.bk.base.datahub.databus.pipe.cal.Cal;
import com.tencent.bk.base.datahub.databus.pipe.convert.Convert;
import com.tencent.bk.base.datahub.databus.pipe.dispatch.Expr;
import com.tencent.bk.base.datahub.databus.pipe.exception.TimeFormatError;
import com.tencent.bk.base.datahub.databus.pipe.record.Field;
import com.tencent.bk.base.datahub.databus.pipe.record.Fields;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Stack;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Context {

    public String origData;

    private static final Logger log = LoggerFactory.getLogger(Context.class);
    private Fields schema;
    private List<Object> values;
    private boolean isVerifyConf = false;
    private EtlParseDetails etlParseDetails = new EtlParseDetails();
    public Map<String, String> ccCacheConfig = null;


    /**
     * 解析错误的日志.
     */
    public List<FailedValue> badValues;
    Stack<Fields> schemaStack;
    Stack<List<Object>> valuesStack;

    // 平坦化的数据条数
    private int totalMsgNum;
    // 成功平坦化的数据条数
    private int successMsgNum;
    // 数据切分的计数
    private int splitMsgNum;

    public ETLImpl etl;
    public Map<String, Integer> ignoreFields;

    private void init() {
        schema = new Fields();
        schemaStack = new Stack<>();
        valuesStack = new Stack<>();
        badValues = new ArrayList<>();
        ignoreFields = new HashMap<>();
        clearValues();
    }

    public Context() {
        init();
    }

    public Context(ETLImpl etl) {
        this.etl = etl;
        init();
    }

    /**
     * 扩充解析失败列表.
     */
    public void appendBadValues(Object data, RuntimeException e) {
        FailedValue item = new FailedValue(data, e);
        badValues.add(item);
    }

    /**
     * 扩充解析失败列表.
     */
    public void appendBadValues(Object data, RuntimeException e, boolean allFailed) {
        if (allFailed) {
            successMsgNum = 0;
        }
        appendBadValues(data, e);
    }


    /**
     * 重置context中收集的数据.
     */
    public void clearValues() {
        values = new ArrayList<Object>();
        badValues = new ArrayList<FailedValue>();
        totalMsgNum = 0;
        successMsgNum = 0;
        splitMsgNum = 0;
    }

    /**
     * 遇到Fields类型的字段需要将当前的schema推到栈中, 采用新的子schema.
     */
    public void pushSchemaStack() {
        schemaStack.push(schema);
        Fields subSchema = new Fields();
        schema.append(subSchema);
        schema = subSchema;
    }

    /**
     * 取得上层的schema.
     */
    public void popSchemaStack() {
        schema = schemaStack.pop();
    }

    /**
     * 数据压栈.
     */
    public void pushValuesStack() {
        valuesStack.push(values);
        List<Object> subValues = new ArrayList<Object>();
        values.add((Object) subValues);
        values = subValues;
    }

    /**
     * 数据压栈.
     */
    public void pushValuesStack(Fields schema) {
        valuesStack.push(values);
        List<Object> subValues = new ArrayList<Object>();
        setValue(schema, subValues);
        values = subValues;
    }

    /**
     * 再填充值的之后会根据schema中的字段信息来填充(主要是map的提取). 这里需要保证list已经初始化好足够的位置
     */
    public void ensureValuesCapacity(int minimal) {
        int more = minimal - values.size();
        while (more-- > 0) {
            values.add(null);
        }
    }

    public void popValuesStack() {
        values = valuesStack.pop();
    }

    /**
     * 获取schema.
     *
     * @return the schema
     */
    public Fields getSchema() {
        return schema;
    }

    /**
     * 设置schema.
     *
     * @param schema the schema to set
     */
    public void setSchema(Fields schema) {
        this.schema = schema;
    }

    /**
     * 获取清洗后的数据.
     *
     * @return the values
     */
    public List<Object> getValues() {
        return values;
    }

    /**
     * 设置值.
     *
     * @param values the values to set
     */
    public void setValues(List<Object> values) {
        this.values = values;
    }

    /**
     * 增长消息总条数计数.
     */
    public void incrTotalMsgNum() {
        this.totalMsgNum += 1;
    }

    public void incrTotalMsgNum(int num) {
        this.totalMsgNum += num;
    }


    /**
     * 增长成功请你先消息计数+1.
     */
    public void incrSuccessMsgNum() {
        this.successMsgNum += 1;
    }

    public void incrSuccessMsgNum(int num) {
        this.successMsgNum += num;
    }


    /**
     * 根据字段对象设置在values中的值.
     */
    public void setValue(Field field, Object value) {
        int index = getSchema().fieldIndex(field);
        ensureValuesCapacity(index + 1);
        values.set(index, value);
    }

    /**
     * 根据字段对象设置在values中的值.
     */
    public void setValue(Fields fields, Object value) {
        int index = getSchema().fieldIndex(fields);
        ensureValuesCapacity(index + 1);
        values.set(index, value);
    }

    /**
     * 平坦化结构字段信息.
     */
    public List<Field> flattenSchema(Fields schema) {
        List<Field> joined = new ArrayList<Field>();
        int innerIdx = -1;

        for (int i = 0; i < schema.size(); i++) {
            if (schema.get(i) instanceof Fields) {
                innerIdx = i;
            } else {
                Field field = (Field) schema.get(i);
                joined.add(field);
            }
        }

        if (innerIdx != -1) {
            joined.addAll(flattenSchema((Fields) schema.get(innerIdx)));
        }

        return joined;
    }

    /**
     * 平坦化收集的值信息.
     */
    @SuppressWarnings("unchecked")
    public List<List<Object>> flattenValues(Fields schema, List<Object> values) {

        List<List<Object>> joined = new ArrayList<List<Object>>();
        List<Object> sampleValues = new ArrayList<Object>();
        int innerIdx = -1;

        if (values.size() != schema.size()) {
            return joined;
        }

        //找到嵌套的迭代器， 当时我为什么要这么做？
        for (int i = 0; i < schema.size(); i++) {
            if (schema.get(i) instanceof Fields) {
                innerIdx = i; //难道这里写是假设数据中只有一个嵌套的迭代么？
            } else {
                sampleValues.add(values.get(i));
            }
        }

        if (innerIdx != -1) {

            if (values.get(innerIdx) == null) {
                int dataId = etl == null ? 0 : etl.getDataId();
                log.debug("{} null context: {} {}, orig data: {}", dataId, values, schema, origData);
                return joined;
            }

            List<List<Object>> complex = new ArrayList<List<Object>>();
            for (List<Object> oneField : (List<List<Object>>) values.get(innerIdx)) {
                complex.addAll(flattenValues((Fields) schema.get(innerIdx), oneField));
            }

            //sample * complex 将嵌套的迭代和外层简单的数据进行笛卡尔积计算
            for (List<Object> oneList : complex) {
                List<Object> collector = new ArrayList<Object>();

                for (Object sampleItem : sampleValues) {
                    collector.add(sampleItem);
                }
                for (Object complexItem : oneList) {
                    collector.add(complexItem);
                }

                joined.add(collector);
            }
        } else {
            joined.add(sampleValues);
        }

        return joined;
    }

    /**
     * 创建ETLResult对象，返回ETL的结果.
     */
    @SuppressWarnings("unchecked")
    public ETLResult toETLResult(ETLImpl etl) {

        Map<Integer, Object> mvalue = new HashMap<Integer, Object>();

        if (etl.getDerive() != null) {
            etl.getDerive().appendPlaceHolder(this);
        }

        // 数据计算
        if (etl.getCals() != null) {
            for (Cal cal : etl.getCals()) {
                cal.appendPlaceHolder(this, null);
            }
        }

        List<List<Object>> flattenedValue = flattenValues(getSchema(), getValues());

        if (etl.timeFieldIdx != -1) {
            flattenedValue = handleTimeField(flattenedValue);
        }

        // 数据转换
        if (etl.converts != null) {
            for (List<Object> oneValue : flattenedValue) {
                for (Convert convert : etl.converts) {
                    if (null != convert && null != oneValue) {
                        oneValue = convert.execute(etl, oneValue);
                    }
                }
            }
        }

        // 数据计算
        if (etl.getCals() != null) {
            for (Cal cal : etl.getCals()) {
                flattenedValue = cal.execute(flattenedValue, etl.getSchema());
            }
        }

        // 数据派生
        if (etl.getDerive() != null) {
            List<List<Object>> newFlattenedValue = new ArrayList<List<Object>>();
            for (List<Object> oneValue : flattenedValue) {
                newFlattenedValue.addAll(etl.getDerive().execute(oneValue, this));
            }
            flattenedValue = newFlattenedValue;
        }

        // 数据切分
        if (etl.dispatch == null) {
            mvalue.put(etl.getDataId(), flattenedValue);
        } else {
            mvalue.put(etl.getDataId(), flattenedValue);
            this.handleDispatch(flattenedValue, mvalue);
        }

        return new ETLResult(etl, mvalue, badValues, totalMsgNum, successMsgNum, splitMsgNum);
    }

    private void handleDispatch(List<List<Object>> flattenedValue, Map<Integer, Object> mvalue) {
        for (List<Object> oneValue : flattenedValue) {

            for (Integer dataId : etl.dispatch.keySet()) {

                Expr expr = etl.dispatch.get(dataId);
                if (expr.execute(etl, oneValue)) {
                    splitMsgNum++;
                    List<List<Object>> l = (List<List<Object>>) mvalue.get(dataId);
                    if (mvalue.get(dataId) == null) {
                        l = new ArrayList<>();
                    }
                    l.add(oneValue);
                    mvalue.put(dataId, l);
                }
            }
        }
    }

    private List<List<Object>> handleTimeField(List<List<Object>> flattenedValue) {
        int idx = 0;
        List<Integer> timeFieldIdxes = new ArrayList<Integer>(); //标记数据中时间解析失败的index

        for (List<Object> oneValue : flattenedValue) {
            try {
                changeDtEventTime(etl, oneValue);

                if (successMsgNum == 0) { // 现在只有遍历的时候加这个，如果数据没有遍历，+1
                    incrTotalMsgNum();
                    incrSuccessMsgNum();
                }

            } catch (TimeFormatError e) {
                this.appendBadValues(oneValue, e);
                successMsgNum--;
                timeFieldIdxes.add(idx);
            }
            idx++;
        }

        // 从列表中剔除时间解析失败的数据
        if (!timeFieldIdxes.isEmpty()) {
            List<List<Object>> tmpFlattenedValue = new ArrayList<List<Object>>();
            for (int i = 0; i < flattenedValue.size(); i++) {
                if (!timeFieldIdxes.contains(i)) {
                    tmpFlattenedValue.add(flattenedValue.get(i));
                }
            }

            flattenedValue = tmpFlattenedValue;
        }

        return flattenedValue;
    }

    /**
     * 将数据时间转换成timestamp.
     */
    private void changeDtEventTime(ETLImpl etl, List<Object> l) throws TimeFormatError {
        // 传入的timeFieldIdx为预设值时, 使用本地时间作为清洗默认时间
        if (etl.timeFieldIdx == EtlConsts.DEFAULT_TIME_FIELD_INDEX) {
            // 使用默认时间时,传入的时间格式为 Unix Time Stamp(milliseconds), 时间长度为13
            Long timestamp = etl.getTimeTransformer().transform(new Date().getTime());
            l.add(timestamp);
        } else {
            // 使用用户定义的时间
            Object data = l.get(etl.timeFieldIdx);
            try {
                Long timestamp = etl.getTimeTransformer().transform(data);
                // 不修改原始数据值，在清洗输出结果的尾部增加用户时间字段转换后的时间戳，然后再添加内部补充字段
                l.add(timestamp);
            } catch (TimeFormatError error) {
                // 记录时间解析失败的相关信息
                addErrorMessage(String.format("%s: %s -> (%s)!", TimeFormatError.class.getSimpleName(), data,
                        etl.getTimeTransformer().getFormat()));
                throw error;
            }
        }
    }

    /**
     * 设置忽略的字段.
     */
    public void setIgnoreFields(List<String> ignoreFields) {

        if (ignoreFields != null) {
            for (String fieldName : ignoreFields) {
                this.ignoreFields.put(fieldName, 1);
            }
        }
    }

    /**
     * 获取验证清洗配置的结果
     *
     * @return 清洗验证结果
     */
    public Map<String, Object> getVerifyResult() {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put(EtlConsts.NODES, etlParseDetails.getNodeDetails());
        result.put(EtlConsts.ERRORS, etlParseDetails.getErrors());
        result.put(EtlConsts.ERROR_MESSAGE, etlParseDetails.getErrorMessage());
        result.put(EtlConsts.OUTPUT_TYPE, etlParseDetails.getNodeOutputType());

        return result;
    }

    /**
     * 设置context为验证清洗配置模式
     */
    public void setContextToVerifyConf() {
        isVerifyConf = true;
    }

    public boolean getVerifyConf() {
        return isVerifyConf;
    }

    /**
     * 根据每个节点的label设置其ETL解析结果
     *
     * @param nodeLabel ETL中节点的label
     * @param result 节点ETL的解析结果
     */
    public void setNodeDetail(String nodeLabel, Object result) {
        if (isVerifyConf) {
            etlParseDetails.setNodeDetail(nodeLabel, result);
        }
    }

    /**
     * 设置节点的错误提示信息
     *
     * @param nodeLabel 节点标签
     * @param errorMsg 节点的错误消息
     */
    public void setNodeError(String nodeLabel, String errorMsg) {
        if (isVerifyConf) {
            etlParseDetails.setNodeErrorMsg(nodeLabel, errorMsg);
        }
    }

    /**
     * 设置时间解析失败的错误提示信息
     *
     * @param errMsg 错误提示信息
     */
    private void addErrorMessage(String errMsg) {
        if (isVerifyConf) {
            etlParseDetails.addErrorMessage(errMsg);
        }
    }

    /**
     * 设置节点的输出数据的类型
     *
     * @param nodeLable 节点标签
     * @param type 节点输出的数据的类型
     * @param value 节点输出的数据
     */
    public void setNodeOutput(String nodeLable, String type, Object value) {
        if (isVerifyConf) {
            etlParseDetails.setNodeOutput(nodeLable, type, value);
        }
    }
}

File: src/api/datamanage/utils/rollback/context.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making BK-BASE 蓝鲸基础平台 available.
Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved.
BK-BASE 蓝鲸基础平台 is licensed under the MIT License.
License for BK-BASE 蓝鲸基础平台:
--------------------------------------------------------------------
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
documentation files (the "Software"), to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial
portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""


import uuid
from contextlib import contextmanager
from functools import wraps

from django.core.cache import cache
from django.utils.encoding import force_text
from django.db import models
from django.db.models.signals import pre_save, pre_delete

from common.log import logger

from datamanage import exceptions as dm_errors


CACHE_KEY_TEMPLATE = 'data_model_application_rollback_{}'


class _ContextWrapper(object):
    def __init__(self, func, args):
        self._func = func
        self._args = args
        self._context = func(*args)

    def __enter__(self):
        return self._context.__enter__()

    def __exit__(self, exc_type, exc_value, traceback):
        return self._context.__exit__(exc_type, exc_value, traceback)

    def __call__(self, func):
        @wraps(func)
        def decorator(*args, **kwargs):
            with self._func(*self._args):
                return func(*args, **kwargs)

        return decorator


def create_temporary_reversion(watching_models, expired_time=3):
    """
    自动进行元数据同步的事务上下文管理器，不支持批量update的操作

    :param using: 需要做事务处理的数据库的别名，如Django默认的数据库别名为default
    :param manage_manually: 默认为False，是否需要手动提交事务，可以参考django原生的transaction.atomic()，不推荐使用

    Examples:
        with auto_meta_sync(using='bkdata_basic'):
            DemoModel.objects.create(param1=1, params2='abc')
    """
    return _ContextWrapper(_create_rollback_context, (watching_models, expired_time))


@contextmanager
def _create_rollback_context(watching_models, expired_time=3):
    """
    创建监听需要回滚的Django Model的上下文管理，在退出watch时把记录的所有变更都保存到django cache中，并配置一定的超时时间，以供
    调用方增删改时失败了需要进行回滚

    :param watching_models: 需要进行监听的Django Model列表
    :param expired_time: 待回滚目标过期时间
    """
    rollback_id = uuid.uuid4().hex

    model_managers = []
    for watching_model in watching_models:
        model_manager = RollbackModelManager(watching_model)
        model_manager.watch()
        model_managers.append(model_manager)

    yield rollback_id

    cache_content = {}
    for model_manager in model_managers:
        model_manager.unwatch()
        cache_content[model_manager.model._meta.model_name] = model_manager.cache_content()

    cache_key = CACHE_KEY_TEMPLATE.format(rollback_id)
    cache.set(cache_key, cache_content, expired_time)


def rollback_by_id(rollback_id):
    cache_key = CACHE_KEY_TEMPLATE.format(rollback_id)
    cache_content = cache.get(cache_key)

    if not cache_content:
        raise dm_errors.RollbackContentExpiredError()

    try:
        for model_name, model_cache_content in list(cache_content.items()):
            for primary_key_value, instance in list(model_cache_content['created'].items()):
                instance.delete()

            for primary_key_value, instance in list(model_cache_content['updated'].items()):
                instance.save()

            for primary_key_value, instance in list(model_cache_content['deleted'].items()):
                instance.delete()
    except Exception as e:
        logger.error('Rollback by {} error: {}'.format(rollback_id, e))
        raise dm_errors.RollbackError()


class RollbackModelManager(object):
    """管理待回滚Model"""

    def __init__(self, model):
        self._model = model

        self._created_models = {}
        self._updated_models = {}
        self._deleted_models = {}

    @property
    def model(self):
        return self._model

    def cache_content(self):
        """生成记录于cache用于回滚的内容"""
        return {
            'created': self._created_models,
            'updated': self._updated_models,
            'deleted': self._deleted_models,
        }

    def watch(self):
        if not issubclass(self.model, models.Model):
            raise dm_errors.OnlySupportModelRollbackError()

        for signal, signal_receiver in self.get_signals_and_receivers(self.model):
            signal.connect(signal_receiver, sender=self.model)

    def unwatch(self):
        for signal, signal_receiver in self.get_signals_and_receivers(self.model):
            signal.disconnect(signal_receiver, sender=self.model)

    def get_signals_and_receivers(self, model):
        yield pre_save, self.pre_save_receiver
        yield pre_delete, self.pre_delete_receiver

    def pre_save_receiver(self, sender, instance, **kwargs):
        primary_key_value = force_text(instance.pk)
        # 通过是否有已经生成的primary key来区分当前要保存的instance是需要创建还是需要更新
        if not instance._state.adding:
            if primary_key_value not in self._created_models and primary_key_value not in self._updated_models:
                self._updated_models[primary_key_value] = instance
        else:
            if primary_key_value not in self._created_models:
                self._created_models[primary_key_value] = instance

    def pre_delete_receiver(self, sender, instance, **kwargs):
        primary_key_value = force_text(instance.pk)
        if primary_key_value not in self._deleted_models:
            self._deleted_models[primary_key_value] = instance
